### ISSUE TITLE
 bundle failed: 'subDays' is not exported by node_modules\date-fns\index.js [SOLVED]

### ISSUE THREAD

**Issue by dickyjayaumbara** [date-fns/date-fns#242]
hey guys, i have issue with date-fns when, iam using on ionic v2, can someowne help me?

this complete message when iam build app

[03:42:04]  bundle failed: 'subDays' is not exported by node_modules\date-fns\index.js (imported by src\pages\calendar\calendar.ts).

```
        For help fixing this error see https://github.com/rollup/rollup/wiki/Troubleshooting#name-is-not-exported-by-module
```

i try to rollup, but iam very confused whit that, help me

**Comment by kossnocorp** [comment]
Hey, which npm version you use?

**Comment by ndrd** [comment]
Same problem here  ☹️

**Comment by kossnocorp** [comment]
@ndrd can you tell more about your setup? Node & npm versions, library versions, etc. If you can, please provide a repo that I can close and debug the error locally.

**Comment by umidbekkarimov** [comment]
@dickyjayaumbara, @ndrd 

So I guess you're requesting it as `import {subDays} from 'date-fns'` and rollup doesn't understands how to map commonjs files? Need your `rollup.config.js` files for example.

@kossnocorp I think src also should be migrated from commonjs to modules (https://lebab.io/ will help)

**Comment by havenchyk** [comment]
@umidbekkarimov could you please clarify a little bit more, what's the problem with rollup's understanding of commonjs?

**Comment by dickyjayaumbara** [comment]
@umidbekkarimov thanks for your help, its make me check the rollup.config.js, and iam try to edit that file, and now the problem is gone.

for @ndrd u can try my solution too,
- go to node_modules\@ionic\app-scripts\config
- edit your rollup.config.js
- search "namedExport" and add the module like this
  namedExports: {
          'node_modules/date-fns/index.js': [ "addDays",
          "addHours",
          "addISOYears",
          "addMilliseconds",
          ...
          ... //add the module u wanna exported
          ...
          "subWeeks",
          "subYears" ]
        }

**Comment by umidbekkarimov** [comment]
@dickyjayaumbara you're welcome 👍  

@havenchyk i'm note rollup user, but AFAIK it doesn't work with common js syntax by default and you need to configure [rollup-plugin-commonjs](https://github.com/rollup/rollup-plugin-commonjs) for it, and in case of `date-fns` lib you also have to configure [custom named exports](https://github.com/rollup/rollup-plugin-commonjs#custom-named-exports) (you can see in @dickyjayaumbara example)

Long story short: 

```javascript
rollup({
  plugins: [
    commonjs({
      namedExports: {
        'node_modules/date-fns/index.js': [/* exports you need */]
      }
    })
  ]
})
```
Also you can use `Object.keys(require('date-fns'))` to obtain all exports from `date-fns`

**Comment by havenchyk** [comment]
@dickyjayaumbara well, it looks like it's not a problem with `date-fns`, but with rollup's way to handle imports. 

So the issue can be closed, feel free to reopen anytime if you would like to add some notes to the readme how to solve this issue.

**Comment by kossnocorp** [comment]
@umidbekkarimov please tell me if it's possible to improve `rollup` support. Thanks everyone!

**Comment by umidbekkarimov** [comment]
I think it should be issue in  [rollup-plugin-commonjs](https://github.com/rollup/rollup-plugin-commonjs) itself (as i said before i'm not rollup user (go team webpack)), cause date-fns follows syntax that they described here [custom-named-exports](https://github.com/rollup/rollup-plugin-commonjs#custom-named-exports). 
Maybe changing syntax of index file to
```
exports.foo = require('./foo/index.js')
exports.bar = require('./bar/index.js')
```
can be the workaround for rollup users

**Comment by lazarljubenovic** [comment]
The real solution to this is to use ES6 modules.
